File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/device/sharedpreferences/UserStorage.java
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.device.sharedpreferences;

import android.content.SharedPreferences;

import com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.data.model.UserEntity;

import io.reactivex.Completable;
import io.reactivex.Single;

/**
 * Created by MateuszW on 2018-03-31.
 */
public class UserStorage {

    private static final String SESSION_TOKEN = "sessionToken";
    private static final String USERNAME = "username";
    private static final String EMAIL = "email";
    private static final String FIRST_NAME = "firstName";
    private static final String LAST_NAME = "lastName";
    private static final String USER_ID = "userId";
    private final SharedPreferences preferences;

    public UserStorage(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    public void saveUser(UserEntity user) {
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString(SESSION_TOKEN, user.getSessionToken());
        editor.putString(USERNAME, user.getUsername());
        editor.putString(EMAIL, user.getEmail());
        editor.putString(FIRST_NAME, user.getFirstName());
        editor.putString(LAST_NAME, user.getLastName());
        editor.putString(USER_ID, user.getUserId());
        editor.apply();
    }

    public UserEntity getUser() {
        final UserEntity userEntity = new UserEntity(preferences.getString(USER_ID, ""));
        userEntity.setSessionToken(preferences.getString(SESSION_TOKEN, ""));
        userEntity.setUsername(preferences.getString(USERNAME, ""));
        userEntity.setEmail(preferences.getString(EMAIL, ""));
        userEntity.setFirstName(preferences.getString(FIRST_NAME, ""));
        userEntity.setLastName(preferences.getString(LAST_NAME, ""));
        return userEntity;
    }

    public Single<Boolean> isUserLogin() {
        return Single.defer(() -> Single.just(!preferences.getString(SESSION_TOKEN, "").isEmpty()));
    }

    public Completable logout() {
        SharedPreferences.Editor editor = preferences.edit();
        editor.clear();
        editor.apply();
        return Completable.complete();
    }

    public String getEmail() {
        return preferences.getString(EMAIL, "");
    }

    public String getUserId() {
        return preferences.getString(USER_ID, "");
    }

    public String getToken() {
        return preferences.getString(SESSION_TOKEN, "");
    }
}

File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/device/sharedpreferences/UserStorage.kt
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.device.sharedpreferences

import android.content.SharedPreferences
import com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.domain.model.User
import io.reactivex.Completable
import io.reactivex.Single

class UserStorage(private val preferences: SharedPreferences) {

    val user: User
        get() {
            return User(
                preferences.getString(USER_ID, "")
                , preferences.getString(SESSION_TOKEN, "")
                , preferences.getString(USERNAME, "")
                , preferences.getString(EMAIL, "")
                , preferences.getString(FIRST_NAME, "")
                , preferences.getString(LAST_NAME, "")
            )
        }

    val isUserLogin: Single<Boolean>
        get() = Single.defer { Single.just(!preferences.getString(SESSION_TOKEN, "")!!.isEmpty()) }

    val email: String?
        get() = preferences.getString(EMAIL, "")

    val userId: String?
        get() = preferences.getString(USER_ID, "")

    val token: String?
        get() = preferences.getString(SESSION_TOKEN, "")

    fun saveUser(user: User) {
        val editor = preferences.edit()
        editor.putString(SESSION_TOKEN, user.sessionToken)
        editor.putString(USERNAME, user.username)
        editor.putString(EMAIL, user.email)
        editor.putString(FIRST_NAME, user.firstName)
        editor.putString(LAST_NAME, user.lastName)
        editor.putString(USER_ID, user.userId)
        editor.apply()
    }

    fun logout(): Completable {
        val editor = preferences.edit()
        editor.clear()
        editor.apply()
        return Completable.complete()
    }

    companion object {
        private const val SESSION_TOKEN = "sessionToken"
        private const val USERNAME = "username"
        private const val EMAIL = "email"
        private const val FIRST_NAME = "firstName"
        private const val LAST_NAME = "lastName"
        private const val USER_ID = "userId"
    }
}
